src/vm/amd64/umthunkstub.S

```diff
@@ -0,0 +1,368 @@
+        jnz             LOCAL_LABEL(UMThunkStub_CopyStackArgs)                                                                   // stack_args
+        
+LOCAL_LABEL(UMThunkStub_ArgumentsSetup):
+        mov             rdi,  [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET +  0h] 
+        mov             rsi,  [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET +  8h] 
+        mov             rdx,   [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 10h] 
+        mov             rcx,   [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 18h]
+        mov             r8,   [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 20h]
+        mov             r9,   [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 28h]
+
+        // @CONSIDER: mark UMEntryThunks that have FP params and only save/restore xmm regs on those calls
+        movdqa          xmm0, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET +  0h]
+        movdqa          xmm1, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 10h]
+        movdqa          xmm2, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 20h]
+        movdqa          xmm3, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 30h]
+        movdqa          xmm4, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 40h]
+        movdqa          xmm5, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 50h]
+        movdqa          xmm6, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 60h]
+        movdqa          xmm7, [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 70h]
+
+        mov             rax, [r11 + OFFSETOF__UMThunkMarshInfo__m_pILStub]                              // rax <- Stub*
+        call            rax
+
+LOCAL_LABEL(PostCall):
+        //
+        // enable preemptive GC
+        //
+        mov             dword ptr [r12 + OFFSETOF__Thread__m_fPreemptiveGCDisabled], 0
+
+        // epilog
+        lea             rsp, [rbp - UMThunkStubAMD64_FRAME_OFFSET + UMThunkStubAMD64_FIXED_STACK_ALLOC_SIZE]
+        pop             rbp                                                                             // stack_args
+        pop             r12
+        ret
+
+
+LOCAL_LABEL(DoThreadSetup):
+        call            C_FUNC(CreateThreadBlockThrow)
+        mov             METHODDESC_REGISTER, [rbp + UMThunkStubAMD64_METHODDESC_SPILL_OFFSET]
+        jmp             LOCAL_LABEL(HaveThread)
+        
+LOCAL_LABEL(DoTrapReturningThreadsTHROW):
+        mov             rdi, r12                                                                  // Thread* pThread
+        mov             rsi, METHODDESC_REGISTER                                                  // UMEntryThunk* pUMEntry
+        call            C_FUNC(UMThunkStubRareDisableWorker)
+        mov             METHODDESC_REGISTER, [rbp + UMThunkStubAMD64_METHODDESC_SPILL_OFFSET]
+
+        jmp             LOCAL_LABEL(InCooperativeMode)
+
+LOCAL_LABEL(UMThunkStub_CopyStackArgs):
+        // rax = cbStackArgs (with 20h for register args subtracted out already)
```

Comment:
"20h for register args subtracted" part of the comment is not true for Unix callconv